### DISCUSSION TITLE
enum type

### DISCUSSION THREAD

**Discussion by michoball** [wanted-9th-3team/frontend-study#43] [프론트 기술 질문 공간]
TypeScript 에서 enum type을 쓰는것에 대해 논쟁이 있는데요. 
enum type의 어떤 특징 때문이지, 그리고 본인의 의견을 말씀해주세요

**Comment by michoball** [comment]
많은 개발자들이 enum 타입의 사용을 지양하자고 하는 이유는 Tree-shaking 이 되지 않기 때문이라고 말합니다. 

tree-shaking 은 js 모듈 번들링 과정에서 사용되지 않는 코드를 제거해서 번들의 크기를 줄여주는 역할을 합니다. 
그런데 typescript 에서 enum으로 사용한 코드는 js로 변환시 IIFE(즉시실행함수)로 구현됩니다. 그리고 이 코드는 tree-shaking과정에서 사용하지 않는 코드로 인식되지 않아 그대로 번들안에 들어가 번들의 크기를 줄이지 못하게 하는 원인이 됩니다. 

해서 enum 타입 사용을 지양하자고 많은 개발자들이 말합니다. 

똑똑한 분들이 쓰지 말자고하시니 안쓰는게 좋을 것 같다는 것이 저의 의견입니다. 그리고 enum은 const enum 이나 as const 로 대체하여 사용이 가능함으로 꼭 쓰고 싶은것이 아니라면 지양하는 것이 좋을 것 같습니다.

**Comment by WooGie911** [comment]
enum 타입의 사용을 지양하는 이유는 Tree-shaking이 되지 않기 때문입니다.

Tree-shaking이란 간단하게 말해 사용하지 않는 코드를 삭제하는 기능을 말합니다. 
나무를 흔들면 죽은 잎사귀들이 떨어지는 모습에서 착안해 Tree-shaking이라고 부르며 
Tree-shaking을 통해 export했지만 import하지 않은 모듈이나 사용하지 않는 코드를 삭제해서 번들 크기를 줄여 페이지가 표시되는 시간을 단축할 수 있습니다. 

enum 타입을 대신해 객체에 as const를 사용하여 union 타입을 추출해 사용하는 방법이 있습니다.
또한 union 추출코드를 제네릭으로 추상화한 아래의 코드로 사용하는 것이 권장됩니다.

물론 임의의 숫자나 문자열을 할당할 수 있으며 하나의 유형으로 사용해서 버그를 줄일 수 있다는 장점이 있지만
Tree-shaking이 되지 않기 때문에 사용을 지양하는것이 좋다고 생각합니다.

**Comment by skyhanull** [comment]
enum 타입의 사용을 지양하는 이유는 Tree-shaking에 따른 문제가 발생하기 때문입니다.
enum과 같은 경우는 Typescript에서 enum이라은 열거형기능을 만들기 위해서 생성해낸 기능으로 다시 js로 트랜스파일이 되면서 
구조가 달라져 rollup과 같은 번들러에서 TreeShaking을 못하게 되는 문제가 있습니다.
 추가적으로 트랜스파일 되면서 즉시실행함수의 형태로 트랜스되고 이는 Rollup에서는 IIFE를 ‘사용하지 않는 코드’라고 판단할 수 없게 됩니다. 그래서 TreeShaking이 안되게 됩니다.
 여기서 트리 쉐이킹은 사용하지 않는 코드를 제거하는 방식입니다.

**Comment by BaekYuri** [comment]
Tree-shaking 문제로 인해 enum 사용을 지양하고 있다고 합니다. 
typescript로 작성 한 후, 다시 Javascript로 트랜스파일이 될 때, rollup과 같은 번들러에서는 enum을 사용하지 않는 코드라고 판단하지 못한다고 합니다. 그렇기 때문에 성능 부분에서 안좋은 면이 있을 수 있어 사용을 지양하고 있습니다.
enum 같은 경우에는 코드의 가독성을 높일 수 있어서 tree-shaking과 같은 문제가 없다면 사용하면 좋을 것이라고 생각하지만, 일단 const enum이나 union과 같은 다른 대체제로 사용할 수 있기 때문에 enum보다는 다른 것을 사용할 것 같습니다.
